File: backend/src/main/scala/no/hamre/polet/service/ProductDataService.scala
package no.hamre.polet.service

import java.time.LocalDate

import no.hamre.polet.dao.Dao
import no.hamre.polet.modell.{Product, ProductLine, ProductRelease}
import no.hamre.polet.parser.FileDownloader
import no.hamre.polet.util.Slf4jLogger

trait ProductDataService {
  def findLatestReleases(): List[Product]

  def updateFromWeb(url: String): DownloadResult

  def findProduct(productId: Long): Option[Product]
  def findAllProduct(): List[Product]
  def findProductByReleaseDate(): List[ProductRelease]
}

class ProductDataServiceImpl(dao: Dao, downloader: FileDownloader) extends ProductDataService with Slf4jLogger {


  override def findLatestReleases(): List[Product] = {
    val products: List[Product] = dao.findLatestReleases
    products.map( p => p.copy(prices = dao.findPrices(p.id)))
  }

  override def updateFromWeb(url: String): DownloadResult = {
    val data = downloader.download(url)
    var whiskies = 0
    var total = 0
    var success = 0
    var failure = 0
    var added = 0
    var priceChanges = 0
    var errors: List[String] = List()
    val lines = data.split("\n")
    log.info(s"${lines.size} lines in file")
    lines.tail
      .map(l => ProductLine(l.split(";")))
      .filter(p => p.varetype.toUpperCase() == "WHISKY")
      .foreach { p => {
      whiskies += 1
      try {
        val stat = update(p)
        success += 1
        if( stat.added ){
          added += 1
        }
        if( stat.priceChanged ){
          priceChanges += 1
        }
        //if( processed > 5) throw StopException("Stopping")
      }catch{
        case e: Exception =>
          log.error(s"Error ${e.getMessage} when parsing: \n$p")
          errors = s"${e.getMessage} : [$p]" :: errors
          failure += 1
      }
    }
    }
    log.info(s"Done downloading and parsing file. ${lines.size} lines read, $whiskies whiskies $failure failed and $success succeeded")
    DownloadResult(lines.size-1, whiskies, success, failure, added, priceChanges, errors)
  }

  case class UpdateStat(added: Boolean, priceChanged: Boolean)

  def update(product: ProductLine): UpdateStat = {
    dao.findByVarenummer(product.varenummer) match {
      case Some(p) =>
        log.info(s"Found varenummer ${p.varenummer} with id ${p.id}")
        dao.updateProductTimestamp(p.id)
        val latestPrice = dao.getLatestPrice(p.id)
        log.info(s"Latest price=$latestPrice")
        if( latestPrice.isEmpty){
          dao.insertPrice(product, p.id, None)
          UpdateStat(added = false, priceChanged = true)
        } else if (latestPrice.get.pris == product.pris) {
          log.info(s"Price unchanged for product ${p.id}")
          UpdateStat(added = false, priceChanged = false)
        } else {
          log.info(s"Price changed from ${latestPrice.get.pris} to ${product.pris} for product ${p.id}")
          dao.priceChanged(latestPrice.get.id, product.datotid, product.pris-latestPrice.get.pris)
          dao.insertPrice(product, p.id, Some(product.pris))
          UpdateStat(added = false, priceChanged = true)
        }
      case None =>
        val productId = dao.insertProduct(product)
        log.info(s"New product inserted. Varenummer ${product.varenummer} got id: $productId")
        dao.insertPrice(product, productId, None)
        UpdateStat(added = true, priceChanged = true)
      case null =>
        throw new RuntimeException(s"findByVarenummer returned null for ${product.varenummer}")
    }
  }

  override def findProduct(id: Long): Option[Product] = {
    val product = dao.findById(id)
    product.map( p => p.copy(prices = dao.findPrices(p.id)))
  }

  override def findProductByReleaseDate(): List[ProductRelease] = {
    val releaseDates: List[LocalDate] = dao.findReleaseDates()
    val miniProducts = dao.findReleasesByDate(releaseDates.head)
    ProductRelease(releaseDates.head, miniProducts, List()) ::
      releaseDates.tail.map( date => ProductRelease(date, List(), List()))
  }

  override def findAllProduct(): List[Product] = {
    dao.findAll.map( p=> p.copy(prices = dao.findPrices(p.id)))
  }
}

case class DownloadResult
(
  total: Int,
  whiskies: Int,
  success: Int,
  failure: Int,
  added: Int,
  priceChanges: Int,
  errors: List[String]
)

File: backend/src/main/kotlin/no/hamre/polet/service/ProductDataService.kt
package no.hamre.polet.service

import no.hamre.polet.dao.Dao
import no.hamre.polet.modell.Whisky
import no.hamre.polet.modell.ProductRelease
import no.hamre.polet.modell.Productline
import no.hamre.polet.vinmonopolet.VinmonopoletClient
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDate


interface ProductDataService {
  fun findLatestReleases(): List<Whisky>

  fun updateFromApi(): Oppdateringsstatus

  fun findWhiskyById(productId: Long): Whisky?
  fun findWhiskies(q: String): List<Whisky>
  fun findAllProduct(): List<Whisky>
  fun findProductByReleaseDate(): List<ProductRelease>
}
@Service
class ProductDataServiceImpl(
    private val dao: Dao,
    private val apiClient: VinmonopoletClient) : ProductDataService {

  private val log = LoggerFactory.getLogger(this.javaClass)

  override fun findLatestReleases(): List<Whisky> {
    val products: List<Whisky> = dao.findBySisteSlipp()
    return products.map { p -> p.copy(prices = dao.findPriser(p.id)) }
  }

  override fun updateFromApi(): Oppdateringsstatus {
    var whiskies = 0
    var success = 0
    var failure = 0
    var added = 0
    var priceChanges = 0
    val errors = mutableListOf<String>()
    var totalProducts = -1
    var start = 0
    val batchSize = 400
    var continueBatch = true
    while (continueBatch) {
      val data = apiClient.doRequest(start, batchSize)
      log.info("${data.whiskies.size} whiskies found in request")
      log.info("Metadata: ${data.metaData}")
      data.whiskies
          //.asSequence()
          .forEach { productline ->
            try {
              val stat = update(productline)
              success += 1
              if (stat.added) {
                added += 1
              }
              if (stat.priceChanged) {
                priceChanges += 1
              }
            } catch (e: Exception) {
              log.error("Error ${e.message} when parsing: \n$productline")
              errors.add("${e.message} : <$productline")
              failure += 1
            }
          }
      log.info("Processed one batch")
      whiskies += data.whiskies.size
      totalProducts = data.metaData.totalCount
      continueBatch = !data.metaData.isDone
      start += batchSize
    }
    log.info("Done downloading from API, $whiskies whiskies $failure failed and $success succeeded")
    return Oppdateringsstatus(
        totalt = totalProducts,
        vellykket = success,
        whiskier = whiskies,
        feil = failure,
        lagtTil = added,
        prisendringer = priceChanges,
        feilmeldinger = errors)
  }

  data class UpdateStat(val added: Boolean, val priceChanged: Boolean)

  fun update(product: Productline): UpdateStat {
    val p = dao.findByVarenummer(product.varenummer)
    if (p != null) {
      log.info("Found varenummer ${p.varenummer} with id ${p.id}")
      dao.setWhiskyUpdated(p.id)
      val latestPrice = dao.findGjeldendePris(p.id)
      log.info("Latest price=$latestPrice")
      if (latestPrice == null) {
        dao.insertPris(product, p.id, null)
        return UpdateStat(added = false, priceChanged = true)
      } else if (latestPrice.pris == product.pris) {
        log.info("Price unchanged for product ${p.id}")
        return UpdateStat(added = false, priceChanged = false)
      } else {
        log.info("Price changed from ${latestPrice.pris} to ${product.pris} for product ${p.id}")
        dao.setPrisendring(latestPrice.id, product.datotid, product.pris - latestPrice.pris)
        dao.insertPris(product, p.id, product.pris)
        return UpdateStat(added = false, priceChanged = true)
      }
    } else { //p == null
      val productId = dao.insertWhisky(product)
      log.info("New product inserted. Varenummer ${product.varenummer} got id: $productId")
      dao.insertPris(product, productId, null)
      return UpdateStat(added = true, priceChanged = true)
    }
  }

  override fun findWhiskyById(productId: Long): Whisky? {
    val product = dao.findById(productId)
    return product?.let { p -> p.copy(prices = dao.findPriser(p.id)) }
  }

  override fun findWhiskies(q: String): List<Whisky> {
    return dao.search(q)
  }

  override fun findProductByReleaseDate(): List<ProductRelease> {
    val releaseDates: List<LocalDate> = dao.findSlippdatoer()
    val miniProducts = dao.findBySlippdato(releaseDates.first())
    return listOf(ProductRelease(releaseDates.first(), miniProducts, listOf())) +
        releaseDates.subList(1, releaseDates.size).map { date -> ProductRelease(date, listOf(), listOf()) }
  }

  override fun findAllProduct(): List<Whisky> {
    return dao.findAll().map { p -> p.copy(prices = dao.findPriser(p.id)) }
  }
}

data class Oppdateringsstatus
(
  val totalt: Int,
  val whiskier: Int,
  val vellykket: Int,
  val feil: Int,
  val lagtTil: Int,
  val prisendringer: Int,
  val feilmeldinger: List<String>
)